allennlp/nn/decoding/decoding_algorithm.py

```diff
@@ -0,0 +1,128 @@
+from allennlp.nn.decoding.decoder_state import DecoderState
+from allennlp.nn import util
+
+
+class DecodingAlgorithm(Registrable):
+    """
+    For state-based models that at every time step take one of several possible actions and then
+    update their state, a ``DecodingAlgorithm`` does two things: (1) it determines how to traverse
+    the exponential state space, and (2) defines a loss function over the result of that traversal.
+
+    Concrete implementations of this abstract base class could do things like maximum marginal
+    likelihood, SEARN, LaSO, or other structured learning algorithms.
+
+    STILL IN PROGRESS.  Right now this is just doing simple maximum likelihood, in a hacky way
+    (because you don't need anything this complex to do maximum likelihood).
+
+    Parameters
+    ----------
+    scheduled_sampling_ratio: float, optional (default = 0.0)
+        At each timestep during training, we sample a random number between 0 and 1, and if it is
+        not less than this value, we use the ground truth labels for the whole batch. Else, we use
+        the predictions from the previous time step for the whole batch. If this value is 0.0
+        (default), this corresponds to teacher forcing, and if it is 1.0, it corresponds to not
+        using target side ground truth labels.  See the following paper for more information:
+        Scheduled Sampling for Sequence Prediction with Recurrent Neural Networks. Bengio et al., 2015.
+    """
+    def __init__(self,
+                 vocab: Vocabulary,
+                 vocab_namespace: str,
+                 scheduled_sampling_ratio: float = 0.0) -> None:
+        self._scheduled_sampling_ratio = scheduled_sampling_ratio
+        # We need the start symbol to provide as the input at the first timestep of decoding, and
+        # end symbol as a way to indicate the end of the decoded sequence.
+        self._start_index = vocab.get_token_index(START_SYMBOL, vocab_namespace)
+        self._end_index = vocab.get_token_index(END_SYMBOL, vocab_namespace)
+
+    def decode(self,
+               num_steps: int,
+               initial_state: DecoderState,
+               decode_step: Callable[[DecoderState, torch.Tensor],
+                                     Tuple[torch.Tensor, List[int], Tuple[torch.Tensor, torch.Tensor]]],
+               training: bool,
+               targets: torch.Tensor = None,
+               target_mask: torch.Tensor = None) -> Dict[str, torch.Tensor]:
+        decoder_state = initial_state
+        finished_states = []
+        for timestep in range(num_steps):
+            decoder_input = self._get_decoder_input(targets,
+                                                    decoder_state.outputs_so_far,
+                                                    timestep,
+                                                    training)
+            action_log_probs, valid_actions, hidden_state = decode_step(decoder_state, decoder_input)
```

Comment:
Why does `decode_step` need to return `valid_actions`? `action_log_probs` should already contain 0 probs for invalid actions.